src/string.c

```diff
@@ -173,197 +163,212 @@ static val_t Str_toString(struct v7 *v7, val_t this_obj, val_t args) {
+          prog == NULL) {
+        throw_exception(v7, "Error", "Invalid String");
+        return V7_UNDEFINED;
       }
-    } while ((arg->fl.fl.re_flags & SLRE_FLAG_G) &&
-             shift < cfa->this_obj->v.str.len);
+    } else
+      prog = ((struct v7_regexp *)v7_to_pointer(ro))->compiled_regexp;
+
+    flag_g = slre_get_flags(prog) & SLRE_FLAG_G;
+    so = to_string(v7, this_obj);
+    s = v7_to_string(v7, &so, &s_len);
+    end = s + s_len;
+
+    do {
+      struct slre_loot sub;
+      struct slre_cap *ptok = sub.caps;
+      int i;
+      if (slre_exec(prog, 0, s, end, &sub)) break;
+      if (arr == V7_NULL) arr = v7_create_array(v7);
+      s = ptok->end;
+      i = 0;
+      do {
+        v7_array_append(v7, arr, v7_create_string(v7, ptok->start,
+                                                  ptok->end - ptok->start, 1));
+        ptok++;
+      } while (!flag_g && ++i < sub.num_captures);
+    } while (flag_g && s < end);
   }
-  if (arr == NULL) TRY(v7_make_and_push(v7, V7_TYPE_NULL));
-  return V7_OK;
-#undef v7
+  return arr;
 }
 
-V7_PRIVATE enum v7_err Str_replace(struct v7_c_func_arg *cfa) {
-#define v7 (cfa->v7) /* Needed for TRY() macro below */
-  struct v7_val *result = v7_push_new_object(v7);
-  const char *out_str;
-  uint8_t own = 1;
-  size_t out_len;
-  int old_sp = v7->sp;
-
-  TRY(check_str_re_conv(v7, &cfa->this_obj, 0));
-  out_str = cfa->this_obj->v.str.buf;
-  out_len = cfa->this_obj->v.str.len;
-  if (cfa->num_args > 1) {
-    const char *const str_end =
-        cfa->this_obj->v.str.buf + cfa->this_obj->v.str.len;
-    char *p = cfa->this_obj->v.str.buf;
+static val_t Str_replace(struct v7 *v7, val_t this_obj, val_t args) {
+  const char *s;
+  size_t s_len;
+  val_t out_str_o;
+  this_obj = to_string(v7, this_obj);
+  s = v7_to_string(v7, &this_obj, &s_len);
+
+  if (s_len != 0 && v7_array_length(v7, args) > 1) {
+    const char *const str_end = s + s_len;
+    char *p = (char *)s;
     uint32_t out_sub_num = 0;
-    struct v7_val *re = cfa->args[0], *str_func = cfa->args[1], *arr = NULL;
+    val_t ro = i_value_of(v7, v7_array_at(v7, args, 0)), str_func = i_value_of(v7, v7_array_at(v7, args, 1));
+    struct slre_prog *prog = NULL;
     struct slre_cap out_sub[V7_RE_MAX_REPL_SUB], *ptok = out_sub;
     struct slre_loot loot;
-    TRY(check_str_re_conv(v7, &re, 1));
-    TRY(regex_check_prog(re));
-    if (v7_is_class(str_func, V7_CLASS_FUNCTION)) {
-      arr = v7_push_new_object(v7);
-      v7_set_class(arr, V7_CLASS_ARRAY);
-      TRY(v7_push(v7, str_func));
-    } else
-      TRY(check_str_re_conv(v7, &str_func, 0));
+    int flag_g = 0;
+
+    if (!v7_is_regexp(ro)) {
+      const char *str;
+      size_t str_len;
+      ro = to_string(v7, ro);
+      str = v7_to_string(v7, &ro, &str_len);
+      if (slre_compile(str, str_len, NULL, 0, &prog, 0) != SLRE_OK ||
+          prog == NULL) {
+        throw_exception(v7, "Error", "Invalid String");
```

Comment:
More than `Error`.
Anyway, not important. I checked the ecmatests, some of them require SyntaxError, some TypeError, we can improve this later.